jobapi/server_test.go

```diff
@@ -0,0 +1,380 @@
+package jobapi_test
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"path"
+	"testing"
+	"time"
+
+	"github.com/buildkite/agent/v3/bootstrap/shell"
+	"github.com/buildkite/agent/v3/env"
+	"github.com/buildkite/agent/v3/jobapi"
+	"github.com/google/go-cmp/cmp"
+)
+
+func pt(s string) *string {
+	return &s
+}
+
+func testEnviron() env.Environment {
+	e := env.Environment{}
+	e.Set("MOUNTAIN", "cotopaxi")
+	e.Set("CAPITAL", "quito")
+
+	return e
+}
+
+func testServer(t *testing.T, e env.Environment) (*jobapi.Server, string, error) {
+	sockName, err := jobapi.NewSocketPath()
+	if err != nil {
+		return nil, "", fmt.Errorf("creating socket path: %w", err)
+	}
+
+	return jobapi.NewServer(shell.TestingLogger{T: t}, sockName, e)
+}
+
+func testSocketClient(socketPath string) *http.Client {
+	return &http.Client{
+		Transport: &http.Transport{
+			DialContext: func(context.Context, string, string) (net.Conn, error) {
+				return net.Dial("unix", socketPath)
+			},
+		},
+	}
+}
+
+func TestServerStartStop(t *testing.T) {
+	t.Parallel()
+
+	env := testEnviron()
+	srv, _, err := testServer(t, env)
+	if err != nil {
+		t.Fatalf("testServer(t, env) error = %v", err)
+	}
+
+	err = srv.Start()
+	if err != nil {
+		t.Fatalf("srv.Start() = %v", err)
+	}
+
+	info, err := os.Stat(srv.SocketPath)
+	if err != nil {
+		t.Fatalf("os.Stat(%q) error = %v", srv.SocketPath, err)
+	}
+
+	isSocket := info.Mode()&os.ModeSocket != 0
```

Comment:
I just discovered https://github.com/golang/go/issues/33357, so this check might need skipping on Windows

Reply:
😭

Reply:
I'm having trouble even calling `os.Stat` on the socket on Windows. But connecting to it is fine.

Reply:
Here's what I ended up with in the client: https://github.com/buildkite/agent/blob/3f9c89d4f519dc89e8ad0b98ee40d1d4bbcd1861/jobapi/client.go#L51-L69